tfhe/src/integer/server_key/radix_parallel/comparison.rs

```diff
@@ -80,46 +117,619 @@ impl ServerKey {
+                u64::from(invert_subtraction_result) ^ overflowed
+            });
+
+            self.key.apply_lookup_table_assign(&mut result_block, &lut);
+
+            BooleanBlock::new_unchecked(result_block)
+        }
+    }
+
+    /// The invert_result boolean is only used when there is one and only one group
+    fn compute_group_borrow_state(
+        &self,
+        invert_result: bool,
+        grouping_size: usize,
+        block_states: Vec<Ciphertext>,
+    ) -> (Vec<Ciphertext>, bool) {
+        if block_states.len() == 1 {
+            return (block_states, true);
+        }
+
+        let message_modulus = self.key.message_modulus.0 as u64;
+        let block_modulus = message_modulus * self.carry_modulus().0 as u64;
+        let num_bits_in_block = block_modulus.ilog2();
+        let num_blocks = block_states.len();
+        let num_groups = num_blocks.div_ceil(grouping_size);
+
+        let num_groupings = num_blocks.div_ceil(grouping_size);
+        let num_carry_to_resolve = num_groupings - 1;
+
+        let sequential_depth = (num_carry_to_resolve as u32) / (grouping_size as u32 - 1);
+
+        let hillis_steel_depth = if num_carry_to_resolve == 0 {
+            0
+        } else {
+            num_carry_to_resolve.ceil_ilog2()
+        };
+
+        let use_sequential_algorithm_to_resolved_grouping_carries = if num_bits_in_block >= 4 {
+            sequential_depth <= hillis_steel_depth
+        } else {
+            true // Hillis-Steele base propagation requires 4 bits
+        };
+
+        // This will be used only if there are at least 2 groups
+        let first_group_propagation_lut = self
+            .key
+            .generate_lookup_table(|block| (block >> (num_bits_in_block as u64 - 1)) & 1);
+
+        // This stores the LUTs that output the propagation result of the other groupings
+        let grouping_chunk_pgn_luts = if use_sequential_algorithm_to_resolved_grouping_carries {
+            // When using the sequential algorithm for the propagation of one grouping to the
+            // other we need to shift the PGN state to the correct position, so we later, when
+            // using them only lwe_add is needed and so noise management is easy
+            //
+            // Also, these LUTs are 'negacylic', they are made to exploit the padding bit
+            // resulting blocks from these LUTs must be added the constant `1 << index`.
+            (0..grouping_size - 1)
+                .map(|i| {
+                    self.key.generate_lookup_table(|block| {
+                        // All bits set to 1 (e.g. 0b1111), means propagate
+                        if block == (block_modulus - 1) {
+                            0
+                        } else {
+                            // u64::MAX is -1 in tow's complement
+                            // We apply the modulus including the padding bit
+                            (u64::MAX << i) % (1 << (num_bits_in_block + 1))
```

Comment:
this has all ones then all zeros in the binary representation right ?

Reply:
This has all bits set to 1

these bits are shited by some amount that is the block position in the group and that position is <= num bits in block. So bits below the block index in group are zeros

Then we apply the modulus 2**(num bits in block+1) which is like doing a bitmask to keep only the bits in range 0..=num bits in block, this means the padding bit is also included